src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb

```diff
@@ -360,17 +369,23 @@ Namespace Microsoft.CodeAnalysis.Semantics
                                 constantValue:=Nothing)
 
                     Dim stepRelationalCode As BinaryOperationKind = Helper.DeriveBinaryOperationKind(BinaryOperatorKind.GreaterThanOrEqual, boundFor.StepValue)
+                    Dim stepConditionIsLifted = stepValue.Type.IsNullableType()
                     Dim stepCondition As IOperation = OperationFactory.CreateBinaryOperatorExpression(stepRelationalCode,
                                  stepValue,
                                  OperationFactory.CreateLiteralExpression(Semantics.Expression.SynthesizeNumeric(stepValue.Type, 0), boundFor.StepValue.Type, boundFor.StepValue.Syntax),
                                  booleanType,
-                                 boundFor.StepValue.Syntax)
+                                 boundFor.StepValue.Syntax,
+                                 stepConditionIsLifted)
+
+                    Dim ifTrueIfFalseIsLifted =
+                        controlVariable.Type.IsNullableType() OrElse
+                        limitValue.Type.IsNullableType()
```

Comment:
Consider moving this before the `If` so it can be shared across both `If` and `Else`.